[GTK] New improved API for page favicons

aperezdc opened the pull request:
#### 07db7fb32f5ce4260e0d9aeea23fee758db81657
<pre>
[GTK] New improved API for page favicons
<a href="https://bugs.webkit.org/show_bug.cgi?id=301086">https://bugs.webkit.org/show_bug.cgi?id=301086</a>

Reviewed by Carlos Garcia Campos.

Add a new WebKitWebView.page-icons property, which holds a WebKitImageList
representing all loaded page icons (favicons) for the currently loaded
content, plus the corresponding webkit_web_view_get_page_icons().

Conversely, add an asynchronous webkit_favicon_database_get_page_icons()
method, which retrieves the WebKitImageList with page icons for a given
URI.

Adding also a WebKitFaviconDatabase:page-icons-changed signal and
possibly a webkit_image_list_find_best() function (or similar), or other
features that build on top of other functionality are out of scope for
this change, which focuses on getting the basics right. Additional
convenience functions may be added later on. The same applies to the
deprecation of the old favicon API, and to enablement of the new page
icons API for the WPE port.

* Source/WebKit/PlatformGTK.cmake: Install WebKitImage.h header.
* Source/WebKit/UIProcess/API/APIIconLoadingClient.h: Add an alternate
interface that can be used to know when all the available page icons
have been processed.
(API::IconLoadingClient::getLoadDecisionForIcons):
(API::IconLoadingClient::iconLoaded):
* Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.cpp: Add new
async API functions to obtain the set of images for a page given its URI.
(webkitFaviconDatabaseGetFaviconInternal):
(webkit_favicon_database_get_page_icons):
(webkit_favicon_database_get_page_icons_finish):
* Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.h.in: Ditto.
* Source/WebKit/UIProcess/API/glib/WebKitIconLoadingClient.cpp: Adapt
to the new, alternative IconLoadingClient interface.
(IconLoadingClient::finishedLoadingIcons):
* Source/WebKit/UIProcess/API/glib/WebKitImage.cpp: Add a boxed type to
represent a set of related images.
(webkitImageListCreate):
(webkit_image_list_ref):
(webkit_image_list_unref):
(webkit_image_list_get_length):
(webkit_image_list_get):
* Source/WebKit/UIProcess/API/glib/WebKitImage.h.in: Ditto.
* Source/WebKit/UIProcess/API/glib/WebKitImagePrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp: Add a new &quot;page-icons&quot;
property that holds a WebKitImageList and is updated once per page load, and
its corresponding getter.
(webkitWebViewUpdatePageIcons):
(webkitWebViewGetProperty):
(webkit_web_view_class_init):
(webkitWebViewGetLoadDecisionForIcons):
(webkitWebViewGetLoadDecisionForIcon): Deleted.
* Source/WebKit/UIProcess/API/glib/WebKitWebView.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitWebViewPrivate.h:
* Source/WebKit/UIProcess/API/glib/webkit.h.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::getLoadDecisionForIcons): Adapt to provide the
load decision for all candidate page icons all at once.
(WebKit::WebPageProxy::getLoadDecisionForIcon): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::getLoadDecisionForIcons): Ditto.
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitFaviconDatabase.cpp:
(serverCallback):
(testFaviconDatabaseGetPageIcons):
(testFaviconDatabaseClear):
(beforeAll):
* Tools/TestWebKitAPI/Tests/WebKitGLib/resources/blue-48x48.png: Added.
* Tools/TestWebKitAPI/Tests/WebKitGLib/resources/red-32x32.png: Added.
* Tools/TestWebKitAPI/glib/TestExpectations.json: The new test typically
passes in an idle machine, but it is still flaky, so mark it as such.

Canonical link: <a href="https://commits.webkit.org/310284@main">https://commits.webkit.org/310284@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8d2ea60f042e6c182ed01e386b0d17798ab3ee93

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/153270 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/26054 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/19686 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/162015 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/106728 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/155143 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/26580 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/26357 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/118497 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/83907 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/156229 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/20740 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/137643 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/99210 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/19817 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/17762 "Passed tests") | [✅ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/9850 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/129464 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/15513 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/164489 "Built successfully") | 
| | [✅ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/7624 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/17107 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/126556 "Passed tests") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/25849 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/21795 "Found 1 new test failure: imported/w3c/web-platform-tests/workers/semantics/structured-clone/shared.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/126714 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/34388 "Built successfully and passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/25851 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/137309 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/82520 "Built successfully") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/21660 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/14087 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/25469 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/89754 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/25168 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/25320 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/25221 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.cpp:
```diff
@@ -278,6 +285,74 @@ cairo_surface_t* webkit_favicon_database_get_favicon_finish(WebKitFaviconDatabas
+
+    auto task = adoptGRef(g_task_new(database, cancellable, callback, userData));
+    database->priv->iconDatabase->loadIconsForPageURL(uriString, IconDatabase::AllowDatabaseWrite::No, [task = WTFMove(task)](Vector<PlatformImagePtr>&& icons) {
+        Vector<GRefPtr<WebKitImage>> images(icons.size(), [&icons](size_t index) {
```
Since we are creating a new Vector from another one I find it easier to read using Vector::map since the lambda receives the item directly instead of an index.

aperezdc replied:
TIL about `Vector::map()`, thanks for the suggestion.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitIconLoadingClient.cpp:
```diff
@@ -36,29 +38,42 @@ class IconLoadingClient : public API::IconLoadingClient {
-        webkitWebViewGetLoadDecisionForIcon(m_webView, icon, [weakWebView = GWeakPtr<WebKitWebView>(m_webView), icon = WTFMove(copiedIcon), completionHandler = WTFMove(completionHandler)] (bool loadIcon) mutable {
-            if (!weakWebView || !loadIcon) {
-                completionHandler(nullptr);
+        m_pendingIcons.clear();
```
Maybe you could use CallbackAggregator for this?

aperezdc replied:
I tried using it when I first wrote this part of the patch; but it didn't work. Unfortunately I cannot remember why right now. Maybe I'll try again and figure out the reason.

aperezdc replied:
Okay, I re-checked this: The callback aggregator would have been useful if we had to call a function for each icon, and in the end do something when all of those async calls end. But here we do a single `webkitWebViewGetLoadDecisionForIcons()` async call, which gets passed the complete icon set, and the result will be the set of load identifiers, also provided all at once &mdash; there is no need to aggregate anything.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -416,6 +418,9 @@ struct _WebKitWebViewPrivate {
     CString faviconURI;
     unsigned long faviconChangedHandlerID;
+#if ENABLE(2022_GLIB_API)
+    GRefPtr<WebKitImageList> pageIcons;
```
WebKitImageList is not a GObject, but a boxed type, to use GRefPtr you need to add a specialization.

aperezdc replied:
Indeed, I was cleaning this up because initially it was a GObject and left a couple of leftovers.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -1207,6 +1224,11 @@ static void webkitWebViewGetProperty(GObject* object, guint propId, GValue* valu
         break;
+#if PLATFORM(GTK) && ENABLE(2022_GLIB_API)
+    case PROP_PAGE_ICONS:
+        g_value_set_object(value, webkit_web_view_get_page_icons(webView));
```
g_value_set_boxed

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -2710,21 +2745,68 @@ void webkitWebViewLoadFailedWithTLSErrors(WebKitWebView* webView, const char* fa
+        }
+
+        CompletionHandler<void(HashSet<WebKit::CallbackID>)> m_completionHandler;
+        HashSet<CallbackID> loadIdentifiers { };
```
```suggestion
        HashSet<CallbackID> loadIdentifiers;
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -2710,21 +2745,68 @@ void webkitWebViewLoadFailedWithTLSErrors(WebKitWebView* webView, const char* fa
+            return;
+        }
+
+        auto webView = adoptGRef(WEBKIT_WEB_VIEW(userData));
```
You should adopt before any early return

aperezdc replied:
Suggestion applied. Indeed we need to do as you say, otherwise thew webview would be leaked in the early return case.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.cpp:
```diff
@@ -263,7 +270,80 @@ cairo_surface_t* webkit_favicon_database_get_favicon_finish(WebKitFaviconDatabas
+    auto task = adoptGRef(g_task_new(database, cancellable, callback, userData));
+    database->priv->iconDatabase->loadIconsForPageURL(uriString, IconDatabase::AllowDatabaseWrite::No, [task = WTF::move(task)](Vector<PlatformImagePtr>&& icons) {
+        Vector<GRefPtr<WebKitImage>> images(icons.size(), [&icons](size_t index) {
+            // TODO: Handle errors, convert to BGRA32 if needed.
```
Use FIXME instead of TODO. Or even better handle the errors and the format conversion :-)

aperezdc replied:
Pixel format conversion implemented.

csaavedra commented on Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.cpp:
```diff
@@ -263,7 +271,94 @@ cairo_surface_t* webkit_favicon_database_get_favicon_finish(WebKitFaviconDatabas
+    g_return_if_fail(pageURI);
+
+    if (!webkitFaviconDatabaseIsOpen(database))
+        return g_task_report_new_error(database, callback, userData, 0, WEBKIT_FAVICON_DATABASE_ERROR, WEBKIT_FAVICON_DATABASE_ERROR_NOT_INITIALIZED, _("Favicons database not initialized yet"));
```
Should it be "Favicon database ..." (singular).

aperezdc replied:
Yes, it should be. It can be a drive-by fixup.

csaavedra commented on Source/WebKit/UIProcess/API/glib/WebKitIconLoadingClient.cpp:
```diff
@@ -36,29 +38,44 @@ class IconLoadingClient : public API::IconLoadingClient {
+            if (!weakWebView) {
+                completionHandler({ });
                 return;
             }
```
```suggestion
            if (!weakWebView) 
                return completionHandler({ });
```

aperezdc replied:
Will do, thanks!

csaavedra commented on Source/WebKit/UIProcess/API/glib/WebKitImage.cpp:
```diff
@@ -418,4 +418,119 @@ static void webkit_image_gloadable_icon_interface_init(GLoadableIconIface* iface
+
+struct _WebKitImageList {
+    Vector<GRefPtr<WebKitImage>> images;
+    int referenceCount { 1 };
```
Not sure if this matters, but why not use `gatomicrefcount`?

aperezdc replied:
There is no code in WebKit using `gatomicrefcount`, so I think it's better to keep doing the same. It's not really an issue because its definition is:

```
% rg gatomicrefcount /usr/include/glib-2.0/glib/gtypes.h
587:typedef gint gatomicrefcount;  /* should be accessed only using atomics */
%
```

If we want to change boxed types to use `gatomicrefcount`, we can do it in a separate patch.

csaavedra reviewed (commented):
Just a couple of minor comments. Feel free to disregard if unimportant.

aperezdc committed:
[GTK] New improved API for page favicons
https://bugs.webkit.org/show_bug.cgi?id=301086

Reviewed by Carlos Garcia Campos.

Add a new WebKitWebView.page-icons property, which holds a WebKitImageList
representing all loaded page icons (favicons) for the currently loaded
content, plus the corresponding webkit_web_view_get_page_icons().

Conversely, add an asynchronous webkit_favicon_database_get_page_icons()
method, which retrieves the WebKitImageList with page icons for a given
URI.

Adding also a WebKitFaviconDatabase:page-icons-changed signal and
possibly a webkit_image_list_find_best() function (or similar), or other
features that build on top of other functionality are out of scope for
this change, which focuses on getting the basics right. Additional
convenience functions may be added later on. The same applies to the
deprecation of the old favicon API, and to enablement of the new page
icons API for the WPE port.

* Source/WebKit/PlatformGTK.cmake: Install WebKitImage.h header.
* Source/WebKit/UIProcess/API/APIIconLoadingClient.h: Add an alternate
interface that can be used to know when all the available page icons
have been processed.
(API::IconLoadingClient::getLoadDecisionForIcons):
(API::IconLoadingClient::iconLoaded):
* Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.cpp: Add new
async API functions to obtain the set of images for a page given its URI.
(webkitFaviconDatabaseGetFaviconInternal):
(webkit_favicon_database_get_page_icons):
(webkit_favicon_database_get_page_icons_finish):
* Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.h.in: Ditto.
* Source/WebKit/UIProcess/API/glib/WebKitIconLoadingClient.cpp: Adapt
to the new, alternative IconLoadingClient interface.
(IconLoadingClient::finishedLoadingIcons):
* Source/WebKit/UIProcess/API/glib/WebKitImage.cpp: Add a boxed type to
represent a set of related images.
(webkitImageListCreate):
(webkit_image_list_ref):
(webkit_image_list_unref):
(webkit_image_list_get_length):
(webkit_image_list_get):
* Source/WebKit/UIProcess/API/glib/WebKitImage.h.in: Ditto.
* Source/WebKit/UIProcess/API/glib/WebKitImagePrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp: Add a new "page-icons"
property that holds a WebKitImageList and is updated once per page load, and
its corresponding getter.
(webkitWebViewUpdatePageIcons):
(webkitWebViewGetProperty):
(webkit_web_view_class_init):
(webkitWebViewGetLoadDecisionForIcons):
(webkitWebViewGetLoadDecisionForIcon): Deleted.
* Source/WebKit/UIProcess/API/glib/WebKitWebView.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitWebViewPrivate.h:
* Source/WebKit/UIProcess/API/glib/webkit.h.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::getLoadDecisionForIcons): Adapt to provide the
load decision for all candidate page icons all at once.
(WebKit::WebPageProxy::getLoadDecisionForIcon): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::getLoadDecisionForIcons): Ditto.
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitFaviconDatabase.cpp:
(serverCallback):
(testFaviconDatabaseGetPageIcons):
(testFaviconDatabaseClear):
(beforeAll):
* Tools/TestWebKitAPI/Tests/WebKitGLib/resources/blue-48x48.png: Added.
* Tools/TestWebKitAPI/Tests/WebKitGLib/resources/red-32x32.png: Added.
* Tools/TestWebKitAPI/glib/TestExpectations.json: The new test typically
passes in an idle machine, but it is still flaky, so mark it as such.

Canonical link: https://commits.webkit.org/310284@main
